sql/core/src/main/scala/org/apache/spark/sql/execution/SQLExecution.scala

```diff
@@ -164,4 +166,19 @@ object SQLExecution {
       }
     }
   }
+
+  /**
+   * Wrap passed function to ensure necessary thread-local variables like
+   * SparkContext local properties are forwarded to execution thread
+   */
+  def withThreadLocalCaptured[T](sparkSession: SparkSession, exec: ExecutionContext)(
+    body: => T): Future[T] = {
+    val activeSession = sparkSession
+    val localProps = Utils.cloneProperties(sparkSession.sparkContext.getLocalProperties)
+    Future {
+      SparkSession.setActiveSession(activeSession)
```

Comment:
Can you please unset/restore both of them to their old state?

Reply:
Do you mean i should clear the thread locals.? as these are worker threads, they do not have any old state as such. For example, when the thread is created first time, it will inherit from parent thread, but when resued, restoring would cause the previous properties to be set which is stale. IMO, setting this on every execution should suffice. Please correct me if wrong

Reply:
I think we should restore them to their previous state. In this particular case you can hold on to a SparkSession that is no longer in use.